interim-jan-2021/resolver-behavior-self-description.md

```diff
@@ -0,0 +1,30 @@
+# Resolver behavior self-description
+
+## Defining local-only namespaces
+* Express namespaces which only this resolver can resolve
+    * Authoritatively if the namespace collides with any global names
+* Ex 1: Enterprise resolvers serving corporation-specific namespaces
+* Ex 2: Public Wi-Fi or cellular networks offering network-local services
+
+## Defining per-namespace optimization
+* Express namespaces for which this resolver provides preferable resolutions
+* Ex 1: ISP routes public content requests to network caches
+* Ex 2: Public resolver designated to serve a namespace to limit parties privy to resolution
+
+## Defining resolver identity
+* Express information consumable by humans describing the resolver's identity
+* Ex 1: Provide human-legible documentation
+    * Most likely a web page link to explain server identity, terms of use, etc.
+	* Not used for decision making by any protocol peer; communicated to clients for display to users
+
+## Defining protocol support
+* Express what optional DNS-related functionality is supported 
+* Ex 1: DNS Extended Errors and which codes to expect
+      * Not exhaustive: server can still return any code
+      * Codes 15-17 indicate kinds of filtering the resolver implements
```

Comment:
Unfortunately the error codes are not complete, the client does not know if the server is performing filtering to block malware or phishing.

Reply:
I don't think it makes sense for the IETF to try to maintain a taxonomy of all of the reasons someone might want to block a domain.  That is a human question, not a technical question, and I think it belongs in the realm of human communication, e.g. on a webpage.  I don't want us to end up debating whether to add "blasphemy" to an IANA registry.  However, this list of requirements is not exhaustive, so it does not preclude the WG from defining such a feature, either now or later (via the extensibility required by point 3).

Reply:
> the client does not know if the server is performing filtering to block malware or phishing.

And I think that's ok. Such details are not actionable by the DNS client anyway. For example:
- Error 15, Blocked: it's the resolver's choice (the error could be bypassed by using another resolver)
- Error 16, Censored: it's a third party's choice (the error probably might not be bypassed by using another resolver)
- Error 17, Filtered: it's the client's choice (the error can be bypassed by updating the resolver with client filtering expectations)

Getting more specific than that doesn't help because it wouldn't let a client take action without human intervention. Today, a client might do something like this:
- 15? Repeat the query to a secondary resolver
- 16? Give up
- 17? Log for local admin to review for action ("oh I didn't mean to filter that", "oh my child is going to get grounded for trying to go to that site", etc.)

Reply:
Agree with both your responses, we proposed https://tools.ietf.org/html/draft-reddy-dnsop-error-page-05 to provide an error page to the user explaining why a DNS query was filtered..

My point is if the DoH server offers multiple DoH services, For example (1) malware blocking and DNSSEC (2) no filtering and no DNSSEC (similar to https://www.quad9.net/doh-quad9-dns-servers/). User has a choice to opt for a) or b) and the user needs to know the above resolver information to make a selection. For instance an home admin could opt for (1) for all the devices attached to the home network.

Reply:
My view on this is that the templates should be communicated without an attempt to describe their behavior differences in the protocol (since it cannot be automatically be confirmed or reasoned on) and let the user opt-in to a choice via the documentation provided. There should be a default "here's how to upgrade path" that requires no choice.

Basically, info beyond EDE should be considered non-verifiable (given it would require much larger constructs such as a trusted auditing authority) and provided via the documentation for optional exploration by interested users. Today, such behavioral differences are usually provided by different IP addresses (different server identities).